realm/realm-library/src/main/java/io/realm/internal/Context.java

```diff
@@ -16,32 +16,60 @@
 
 package io.realm.internal;
 
-import java.lang.ref.Reference;
 import java.lang.ref.ReferenceQueue;
 import java.util.ArrayList;
-import java.util.HashMap;
 import java.util.List;
 
 public class Context {
 
+    // Pool to hold the phantom references.
+    // The size of array for storing phantom references will never decrease. Instead, we use another array to hold the
+    // index of the free slot. When adding the reference, pick the last index from freeIndexList and put the reference
+    // to the corresponding slot.
+    private static class ReferencesPool {
+        ArrayList<NativeObjectReference> arrayList = new ArrayList<>();
+        ArrayList<Integer> freeIndexList = new ArrayList<Integer>();
```

Comment:
Something inside me really dislike using ArrayList/Integer for this as we are wasting quite a lot of memory compared to a `int[]` list. However it would also increase the complexity of this quite a bit.